internal/nfs/controller/volume.go

```diff
@@ -0,0 +1,237 @@
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	fscore "github.com/ceph/ceph-csi/internal/cephfs/core"
+	"github.com/ceph/ceph-csi/internal/cephfs/store"
+	fsutil "github.com/ceph/ceph-csi/internal/cephfs/util"
+	"github.com/ceph/ceph-csi/internal/util"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+)
+
+const (
+	// clusterNameKey is the key in OMAP that contains the name of the
+	// NFS-cluster. It will be prefixed with the journal configuration.
+	clusterNameKey = "nfs.cluster"
+)
+
+type NFSVolume struct {
+	// ctx is the context for this short living volume object
+	ctx context.Context
+
+	volumeID   string
+	clusterID  string
+	mons       string
+	fscID      int64
+	objectUUID string
+
+	// TODO: drop in favor of a go-ceph connection
+	cr        *util.Credentials
+	connected bool
+	conn      *util.ClusterConnection
+}
+
+func NewNFSVolume(ctx context.Context, volumeID string) (*NFSVolume, error) {
+	vi := util.CSIIdentifier{}
+
+	err := vi.DecomposeCSIID(volumeID)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding volume ID (%s): %w", volumeID, err)
+	}
+
+	return &NFSVolume{
+		ctx:        ctx,
+		volumeID:   volumeID,
+		clusterID:  vi.ClusterID,
+		fscID:      vi.LocationID,
+		objectUUID: vi.ObjectUUID,
+		conn:       &util.ClusterConnection{},
+	}, nil
+}
+
+func (nv *NFSVolume) String() string {
+	return nv.volumeID
+}
+
+func (nv *NFSVolume) Connect(cr *util.Credentials) error {
+	if nv.connected {
+		return nil
+	}
+
+	var err error
+	nv.mons, err = util.Mons(util.CsiConfigFile, nv.clusterID)
+	if err != nil {
+		return fmt.Errorf("failed to get MONs for cluster (%s): %w", nv.clusterID, err)
+	}
+
+	err = nv.conn.Connect(nv.mons, cr)
+	if err != nil {
+		return fmt.Errorf("failed to connect to cluster: %w", err)
+	}
+
+	nv.cr = cr
+	nv.connected = true
+
+	return nil
+}
+
+func (nv *NFSVolume) Destroy() {
+	if nv.connected {
+		nv.conn.Destroy()
+		nv.connected = false
+	}
+}
+
+func (nv *NFSVolume) GetExportPath() string {
+	return "/" + nv.volumeID
+}
+
+func (nv *NFSVolume) CreateExport(backend *csi.Volume) error {
+	if !nv.connected {
+		return fmt.Errorf("can not created export for %q: not connected", nv)
+	}
+
+	fs := backend.VolumeContext["fsName"]
+	nfsCluster := backend.VolumeContext["cephNFS"]
+	path := backend.VolumeContext["subvolumePath"]
+
+	err := nv.setNFSCluster(nfsCluster)
+	if err != nil {
+		return fmt.Errorf("failed to set NFS-cluster: %w", err)
+	}
+
+	// ceph nfs export create cephfs ${FS} ${NFS} /${EXPORT} ${SUBVOL_PATH}
+	args := []string{
+		"--id", nv.cr.ID,
+		"--keyfile=" + nv.cr.KeyFile,
+		"-m", nv.mons,
+		"nfs",
+		"export",
+		"create",
+		"cephfs",
+		fs,
+		nfsCluster,
+		nv.GetExportPath(),
+		path,
+	}
+
+	// TODO: use new go-ceph API
+	_, stderr, err := util.ExecCommand(nv.ctx, "ceph", args...)
+	if err != nil {
+		return fmt.Errorf("executing ceph export command failed (%w): %s", err, stderr)
```

Comment:
do we need to handle the already exported errors if we get any? or this call is idempotent?